File: leetcode/530-minimum-absolute-difference-in-bst/main.go
package main

import (
	"fmt"
	"math"
)

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

/*
	1st approach
	- convert bst into a sorted array
	- the min diff is actually the min diff between 2 adjacent elements
	Time	O(2n)
	Space O(n)
*/
func getMinimumDifference(root *TreeNode) int {
	arr := []int{}
	var inorder func(node *TreeNode)
	inorder = func(node *TreeNode) {
		if node == nil {
			return
		}
		inorder(node.Left)
		arr = append(arr, node.Val)
		inorder(node.Right)
	}
	inorder(root)
	min := math.MaxInt64
	for i := 0; i < len(arr)-1; i++ {
		if arr[i+1]-arr[i] < min {
			min = arr[i+1] - arr[i]
		}
	}
	return min
}

func main() {
	// 			7
	//		3		11
	// 0	 6 8  15
	root := &TreeNode{7,
		&TreeNode{3,
			&TreeNode{0,
				nil,
				nil,
			},
			&TreeNode{6,
				nil,
				nil,
			},
		},
		&TreeNode{11,
			&TreeNode{8,
				nil,
				nil,
			},
			&TreeNode{15,
				nil,
				nil,
			},
		},
	}
	fmt.Println(getMinimumDifference(root))
}

File: leetcode/530-minimum-absolute-difference-in-bst/main.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

"""
    3rd: dfs with ranges

    Time    O(N)
    Space   O(H)
     84 ms, faster than 27.50%
"""


class Solution:
    def getMinimumDifference(self, root: TreeNode) -> int:
        self.res = sys.maxsize
        self.dfs(root, -sys.maxsize, sys.maxsize)
        return self.res

    def dfs(self, node, minVal, maxVal):
        if node == None:
            return
        self.res = min(self.res, abs(node.val - minVal),
                       abs(node.val - maxVal))
        self.dfs(node.left, minVal, node.val)
        self.dfs(node.right, node.val, maxVal)
